samples/bluetooth/channel_sounding/src/connected_cs.c

```diff
@@ -0,0 +1,498 @@
+static K_SEM_DEFINE(sem_config_created, 0, 1);
+static K_SEM_DEFINE(sem_cs_security_enabled, 0, 1);
+static K_SEM_DEFINE(sem_procedure_done, 0, 1);
+static K_SEM_DEFINE(sem_connected, 0, 1);
+static K_SEM_DEFINE(sem_discovered, 0, 1);
+static K_SEM_DEFINE(sem_written, 0, 1);
+static K_SEM_DEFINE(sem_data_received, 0, 1);
+
+#define CS_CONFIG_ID      0
+#define NUM_MODE_0_STEPS  1
+#define NAME_LEN          30
+#define STEP_DATA_BUF_LEN 512 /* Maximum GATT characteristic length */
+
+void estimate_distance(uint8_t *local_steps, uint16_t local_steps_len, uint8_t *peer_steps,
+		       uint16_t peer_steps_len, uint8_t n_ap, enum bt_conn_le_cs_role role);
+static ssize_t on_attr_write_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+static uint16_t step_data_attr_handle;
+static struct bt_conn *connection;
+static enum bt_conn_le_cs_role role_selection;
+static uint8_t n_ap;
+static uint8_t latest_num_steps_reported;
+static uint16_t latest_step_data_len;
+static uint8_t latest_local_steps[STEP_DATA_BUF_LEN];
+static uint8_t latest_peer_steps[STEP_DATA_BUF_LEN];
+
+static struct bt_uuid_128 step_data_char_uuid =
+	BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x87654321, 0x4567, 0x2389, 0x1254, 0xf67f9fedcba8));
+static const struct bt_uuid_128 step_data_svc_uuid =
+	BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x87654321, 0x4567, 0x2389, 0x1254, 0xf67f9fedcba9));
+static struct bt_gatt_attr gatt_attributes[] = {
+	BT_GATT_PRIMARY_SERVICE(&step_data_svc_uuid),
+	BT_GATT_CHARACTERISTIC(&step_data_char_uuid.uuid, BT_GATT_CHRC_WRITE,
+			       BT_GATT_PERM_WRITE | BT_GATT_PERM_PREPARE_WRITE, NULL,
+			       on_attr_write_cb, NULL),
+};
+static struct bt_gatt_service step_data_gatt_service = BT_GATT_SERVICE(gatt_attributes);
+
+static ssize_t on_attr_write_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	if (flags & BT_GATT_WRITE_FLAG_PREPARE) {
+		return 0;
+	}
+
+	if (offset) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len != sizeof(latest_local_steps)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	if (flags & BT_GATT_WRITE_FLAG_EXECUTE) {
+		uint8_t *data = (uint8_t *)buf;
+
+		memcpy(latest_peer_steps, &data[offset], len);
+		k_sem_give(&sem_data_received);
+	}
+
+	__ASSERT(role_selection == BT_CONN_LE_CS_ROLE_INITIATOR, "Unexpected GATT write cb");
+
+	return len;
+}
+
+static const char sample_str[] = "CS Sample";
+static const struct bt_data ad[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, "CS Sample", sizeof(sample_str) - 1),
+};
+
+static void subevent_result_cb(struct bt_conn *conn, struct bt_conn_le_cs_subevent_result *result)
+{
+	latest_num_steps_reported = result->header.num_steps_reported;
+	n_ap = result->header.num_antenna_paths;
+
+	if (result->step_data_buf) {
+		if (result->step_data_buf->len <= STEP_DATA_BUF_LEN) {
```

Comment:
Yes, I tried to keep things simple here in zephyr so I used a long GATT write, which only needs one call to bt_gatt_write(). To send more than this I think you have to split up the data and send it over multiple writes or indications, which seemed a bit complex.

For procedures with more than one subevent, you're right that it'll only use the last subevent's results. I agree this isn't so realistic, do you have a suggestion on how to do this better while keeping things relatively simple ?

Reply:
Perhaps one could retrieve and concatenate the data from all subevents in the procedure, before doing the bt_gatt_write(). We would still have the limitation of 512 bytes of data per procedure, so I don't think it makes sense to implement this.

Another option would be to use the first subevent in the procedure instead of the last one. Reasons for doing that instead would be:
- The first subevent in a procedure is less likely to be "dropped". Currently, our controller will drop all subsequent subevents if there is an issue with one of them. For example if it can not be scheduled.
- The last subevent in a procedure often have less data then other subevents, for example if the procedure terminates because of reaching the channel map repetitions, reaching maximum number of steps in a procedure or reaching max_procedure_len